### ISSUE TITLE
Relative redirect_url without secure cookies is broken

### ISSUE THREAD

**Issue by rassie** [oauth2-proxy/oauth2-proxy#1191]
(somewhat connected to https://github.com/oauth2-proxy/oauth2-proxy/issues/1190)

When using relative `--redirect-url` (e.g. `/oauth2/callback`) with `--cookie-secure=false`, the scheme in the redirect URL is not set, so that it's evaluated as relative. With secure cookies, it's set to HTTPS unconditionally.

## Expected Behavior

1. Set relative redirect URL and disable secure cookies
2. Open `oauth2-proxy`-protected site
3. Login at IDP
4. Get redirected to your protected site

## Current Behavior

1. Set relative redirect URL and disable secure cookies
2. Open `oauth2-proxy`-protected site
3. Login at IDP
4. Get redirected to `https://your.idp.url/auth//your.site.url/oauth2/callback`.

## Possible Solution

Set scheme on the redirect URL explicitely if the URL is relative.

**Comment by rassie** [comment]
A workaround proposal: after https://github.com/oauth2-proxy/oauth2-proxy/blob/06808704a3762de73bed1cc92b077420eb93d735/oauthproxy.go#L932-L936 

add the following snippet: 

```
	// If the scheme is not set, it means
	// 
	// a) the redirect URL in the configuration is relative
	// 
	// b) the protected site is HTTP or p.CookieSecure is false
	// 
	// We have to figure out the scheme, since otherwise the redirect URI will
	// be incomplete. Since our redirect URL is relative, it means our Referer
	// will be the same site, so we can take the scheme from there.
	if rd.Scheme == "" {
		url, err := url.Parse(req.Referer())
		if err == nil {
			rd.Scheme = url.Scheme
		}
	}
```

**Comment by rassie** [comment]
OK, this workaround is void -- we are not guaranteed to have a referrer. Back to the drawing board...

**Comment by NickMeves** [comment]
Can you elaborate on how a relative URL would work in your environment?

This is the redirect from the IdP (99.9% of the time a 3rd party domain) that gets passed as a querystring parameter in the `authorization_endpoint` request to the IdP during the first part of OAuth2.

The IdP uses this for the client redirect to the callback URL that does the token request portion of the OAuth2 flow.

So with a redirect triggered by the IdP, a relative URL to `/oauth2/callback` wouldn't work because the callback URL isn't at the IdP.

Unless you are using an internal IdP hosted on the exact same domain I guess? Is that your setup?

**Comment by NickMeves** [comment]
If you don't explicitly set a host, there's a lot of logic to try to figure out hosts & schemes based on request details & headers.

If you are in a proxied environment, where these might come from X-Forwarded-* headers, you need set the `--reverse-proxy` for those to be trusted.

If you look at your traffic, on the initial redirect to your IdP - what is being set in the `redirect_uri` (or something like that, my memory is fuzzy 😅 ) parameter? That is what governs where your IdP will redirect you.

That value is built with the logic here: https://github.com/oauth2-proxy/oauth2-proxy/blob/06808704a3762de73bed1cc92b077420eb93d735/oauthproxy.go#L916

**Comment by rassie** [comment]
That is more or less my problem: if we assume my protected site to be `http://my.site.example.com`, I'm getting either `https://my.site.example.com` with `--cookie-secure` enabled (which is probably expected) or `//my.site.example.com` without secure cookies as `redirect_uri` in the IDP. The latter should probably never happen, no matter what settings you are using.

**Comment by rassie** [comment]
Oh, and I see there is also https://github.com/oauth2-proxy/oauth2-proxy/commit/9dbd5f5afd87c8ffc60c9bb997ad0038c76d39d4, which is probably a fix. I haven't upgraded to the latest version yet :see_no_evil:. Let me check this out, thanks!

**Comment by rassie** [comment]
From my testing, this has indeed been fixed by #1045. Sorry for the trouble!
